Jake Vendingmachine step4

ChocOZerO opened the pull request:
자판기 step4 구현했습니다.

ChocOZerO committed:
[Update]자판기를 참조타입으로 변경 #19

ChocOZerO committed:
[Refactoring]관리자와 구매하는 사용자 기능 분리 #19

VendingMachine을 기본 상태만 남기고 각 기능을 관리자와 구매자로 구분
VendingMachine을 class로 변경해서 참조타입으로 변경하여 사용
user와 admin struct를 만들어 각각 다른 동작을 할 수 있도록 변경

ChocOZerO committed:
[Update]Main에서 기존과 같이 동작하도록 변경 #20

ChocOZerO committed:
[Update]같은 음료를 구분하기 위해 Equatable conform #21

ChocOZerO committed:
[Add]특정 음료를 제거하는 remove메소드 추가 #21

ChocOZerO committed:
[Refactoring]관리자모드와 사용자모드 구분해서 진행하도록 변경 #21

ChocOZerO committed:
[Fix]참조문제로 생긴 버그 수정 #21

ChocOZerO committed:
[FitCodacy]공백정리

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -24,4 +24,3 @@ while true {
         user.use()
     }
```
mode와 비교하는 값 "1"처럼 하드코딩하기 보다는 의미있는 코드로 만드세요.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -12,31 +12,97 @@ typealias Products = Array<Beverage>
+
+struct VendingMachineUser {
+    private var vendingMachine: VendingMachine
+    private(set) var purchaseList: Products = []
```
VendingMachineUser 나 VendingMachineAdmin 객체가 
VendingMachine 객체에 의존하고 있는데 이렇게 두 객체를 모두 써야만 하는걸까요?
예를 들어 VendingMachine 객체는 숨겨놓고 -User와 -Admin 객체만 사용하면 어떨까요?

ChocOZerO replied:
숨겨놓는다는 의미를 잘 모르겠습니다.
지금 소스는 VendingMachine을 다른데서 사용하지 않도록 fileprivate로 속성을 관리하면서 같은 파일에 다른 두 객체를 배치했습니다. 이걸 숨겨놓는다는 의미라고 생각할 수도 있을까요?

godrm replied:
VendingMachineUser나 VendingMachineAdmin 객체 내부에서만 사용하도록 감춘다는 건,
VendingMachine 클래스 존재 자체를 몰라서 객체 인스턴스를 만들 필요가 없도록하는거죠.
AdminController나 UserController에서도 `let vendingMachine: VendingMachine`를 만들지 않도록 하라는거죠.

ChocOZerO replied:
지금 설계된 로직은 자판기가 존재하고 사람이 관리자인지 일반사용자인지에 따라 다른 동작을 취하도록 하는데요, VendingMachine이 main에서 사라지면 같은 자판기를 공유한다는 개념이 틀어지는데 설계를 좀 바꿔봐야될까요?

godrm replied:
VendingMachine이 클래스가 되서 레퍼런스 타입이면 어떨까요?

ChocOZerO replied:
지금 채택하고 있는 방식이 클래스의 레퍼런스 참조를 활용하고 있습니다.
문제는 인스턴스를 생성하고 그 인스턴스가 양쪽에 전달이 되어야 내부에서 활용할 수 있을것 같아 현재와 같은 로직으로 구현되었습니다.

godrm replied:
아.. class로 바꾼 내용을 확인못했네요.

godrm replied:
VendingMachine 객체 의존성을 갖는 객체가 VendingMachine 클래스 타입 자체에 의존적이기 보다는
VendingMachine이 갖고 있어야할 인터페이스에 의존하도록 DIP를 적용해보면 어떨까요?
그러면 단위 테스트를 할때도 꼭 VendingMachine 객체가 아니라 같은 인터페이스를 가지는 객체로 테스트를 할 수 있게됩니다.

godrm reviewed (commented):
객체 관계에 대해 한 번 더 고민해보세요.

ChocOZerO committed:
[Update]자판기 모드에 하드코딩 제거하고 enum활용 #22

godrm commented on VendingMachine/VendingMachine/AdminController.swift:
```diff
@@ -0,0 +1,101 @@
+
+    init(vendingMachine: VendingMachine) {
+        self.vendingMachine = vendingMachine
+        admin = VendingMachineAdmin.init(vendingMachine: vendingMachine)
```
단지 vendingMachine 변수를 VendingMachineAdmin 을 생성하기 위한 거라면 내부에 선언할 필요까지는 없겠네요.

godrm commented on VendingMachine/VendingMachine/UserController.swift:
```diff
@@ -0,0 +1,62 @@
+            case .buyProduct where selector.1 <= buyableProducts.count:
+                user.buy(category: buyableProducts[selector.1-1])
+                print("\(buyableProducts[selector.1-1])를 구매하셨습니다. \(String(describing: inventory[buyableProducts[selector.1-1]]![0].price))원을 차감합니다.")
+            case .purchaseList:
```
selector.1 은 배열의 index를 의미하나요? 튜플에도 의미를 부여하는 건 어떨까요?
이렇게 반복적으로 `selector.1-1` 형태가 반복되는 것을 줄여야 실수를 덜 하게 됩니다.

ChocOZerO committed:
[FitCodacy]클로저 정리

ChocOZerO committed:
[Update]튜플에 label추가해서 활용 #24

ChocOZerO committed:
[Update]불필요한 변수 제거 #23

ChocOZerO committed:
[Refactoring]VendingMachine객체 관계 구조 개선 #23

VendingMachineProtocol을 만들어 각 하위 객체(user, admin)에서
클래스를 직접 소유하지 않고 프로토콜을 소유할 수 있도록 변경

ChocOZerO committed:
[FitCodacy]공백 정리

ChocOZerO commented:
프로토콜을 이용해서 DIP적용해봤습니다.
codacy에 잘 맞으면서 보기좋게 클로저를 사용할 수 있도록 가이드 코드가 있으면 좋을거같아요.. ㅎㅎㅎ

godrm commented on VendingMachine/VendingMachine/UserController.swift:
```diff
@@ -0,0 +1,62 @@
+                continue
+            }
+            selector = (menu: ValidationChecker.VendingMenu.none, contents: 0)
+        }
```
AdminController.manage() 메소드와 
UserController.user() 메소드가 너무 긴 것 같은데... 메소드 분리가 필요하지 않을까요?

ChocOZerO replied:
분리해서 적용했습니다.

godrm commented on VendingMachine/VendingMachineTests/VendingMachineUserTest.swift:
```diff
@@ -0,0 +1,77 @@
+        user.buy(category: "칸타타커피")
+        XCTAssertEqual(user.purchaseList.count, 1)
+    }
+
```
VendingMachineMock 객체를 이용한 테스트 방식은 좋네요.

ChocOZerO replied:
전에 피드백 주신 부분이 생각나서 적용해보았습니다.

godrm commented on VendingMachine/VendingMachine/ValidationChecker.swift:
```diff
@@ -51,10 +57,21 @@ struct ValidationChecker {
+        let selector = Int(input) ?? 0
+        guard selector >= 1 && selector <= 3 else {
+            throw Errors.notInMenu
         }
```
Int 대신 AdminMenu 열거형으로 비교하는 방식은 어떤가요?

ChocOZerO replied:
열거형 비교방식을 고민하다가 더 좋은 방식이 생각나서 수정했습니다.

ChocOZerO committed:
[Update]컨트롤러 메소드 분리 #25

ChocOZerO committed:
[Update]admin menu selector validate 부분 개선 #26

godrm commented:
수고했습니다. 다음 단계 swiftlint 를 적용하고 확인받으세요 👍
